src/libmcrx.c

```diff
@@ -1152,28 +1160,49 @@ MCRX_EXPORT enum mcrx_error_code mcrx_mnatmap_apply(struct mcrx_ctx *ctx,
     if (ret != MCRX_ERR_OK) {
       return ret;
     }
-    // replace the run time mnat context
-    mcrx_mnatmap_unref(ctx->mnat_map);
+    struct mcrx_mnatmap *original_mnat_map = ctx->mnat_map;
     ctx->mnat_map = clone_context;
     struct mcrx_subscription *sub;
     LIST_FOREACH(sub, &ctx->subs_head, sub_entries) {
-      if (sub->joined) {
-        mcrx_subscription_leave(sub);
+      if (sub->state == MCRX_SUBSCRIPTION_STATE_PENDING) {
         mcrx_subscription_join(sub);
-      } else {
+      } else if (sub->state == MCRX_SUBSCRIPTION_STATE_JOINED && original_mnat_map == NULL) {
+        // previous subscription joined without MNAT
         mcrx_subscription_join(sub);
+      } else if (sub->state == MCRX_SUBSCRIPTION_STATE_JOINED) {
+        struct mcrx_mnat_entry *entry = NULL;
+        struct mcrx_source_group_addrs global_address;
+        memset(&global_address, 0, sizeof(global_address));
+        global_address.addr_type = sub->input.addr_type;
+        memcpy(&global_address.addrs, &sub->input.addrs,
+            sizeof(global_address.addrs));
+        entry = mcrx_mnatmap_find_entry(ctx->mnat_map, &global_address);
+        if (!mcrx_mnatmap_entry_local_equal(entry, sub->mnat_entry)) {
+          mcrx_subscription_leave(sub);
+          mcrx_subscription_join(sub);
+        }
```

Comment:
This needs an "else { sub->mnat_entry = entry; }" to point the sub into the new mnat_map.  It can't stay pointing at the old one we're about to delete.